model-optimizer/extensions/back/InterpolateReshape.py

```diff
@@ -0,0 +1,150 @@
+                    shape=[1, 7, 60, 160]
+    AFTER:
+                input
+            shape=[1, 3, 30, 40]           input_1
+               |                     shape=[1, 4, 60, 160]
+               |                      /        |
+               |                  ShapeOf      |
+               |                    |          |
+               |               Gather          |
+               |     indices=(2, 3); axis=0    |
+               \                    |          |
+                Interpolate(axes=(2, 3))      |
+            shape=[1, 3, 60, 160]             |
+                        \                   /
+                           Concat(axis=1)
+                        shape=[1, 7, 60, 160]
+
+    """
+    enabled = True
+    graph_condition = [lambda graph: graph.graph['cmd_params'].generate_experimental_IR_V10,
+                       lambda graph: graph.graph['cmd_params'].keep_shape_ops]
+    force_clean_up = True
+
+    @staticmethod
+    def make_interpolate_reshapeable(node):
+        concat = node.out_port(0).get_destination().node
+        output_shape = node.out_port(0).data.get_shape()
+
+        interp_axes = [get_canonical_axis_index(output_shape, axis) for axis in node.axes]
+        concat_axis = get_canonical_axis_index(output_shape, concat.axis)
+        if concat_axis in interp_axes:
+            return
+
+        concat_srcs = [port.get_source() for port in concat.in_ports().values()]
+        non_interp_concat_srcs = [src for src in concat_srcs if src.node.soft_get('type') != 'Interpolate']
+        if len(non_interp_concat_srcs) == 0:
+            return
+
+        graph = node.graph
+        src = non_interp_concat_srcs[0]
+
+        shape = Shape(graph, {'name': src.node.soft_get('name', src.node.id) + '/Shape'}).create_node()
+        shape.in_port(0).connect(src)
+        gather = create_op_with_const_inputs(graph, Gather, {1: np.array(node.axes, dtype=np.int32), 2: int64_array(0)},
+                                             {'name': shape.name + '/Gathered'}, shape)
+        node.in_port(1).get_connection().set_source(gather.out_port(0))
+
+    def find_and_replace_pattern(self, graph: Graph):
+        for node in graph.get_op_nodes(type='Interpolate'):
+            if node.in_port(1).get_source().node.soft_get('type') != 'Const':
+                continue
+            dsts = node.out_port(0).get_destinations()
+            if len(dsts) == 1 and dsts[0].node.soft_get('type') == 'Concat':
+                self.make_interpolate_reshapeable(node)
+
+
+class InterpolateReshapeWA(BackReplacementPattern):
+    """
+    Replaces hard-coded 1-port input of Interpolate with reshape-able sub-graph.
+    WARNING: Could cause troubles if model has hard-coded Interpolate intentionally -- rare situation
+
+    BEFORE:
+        input                   Const
+    shape=[1, 3, 30, 40]      value=[60, 160]
+            \                   /
+           Interpolate(axes=(2, 3))
+            shape=[1, 3, 60, 160]
```

Comment:
Our observation shows that 99.99% models were designed with intention for this Interpolate to be reshape-able -- to really change the output shape with respect to new input shape.

Reply:
Also, I would like to highlight once more, that this transformation if disabled. It will be available for user that would like to fix this original model non-reshape-ability